### ISSUE TITLE
PMVComfRecipe+IndoorComfAnalysis annual simulation

### ISSUE THREAD

**Issue by ayezioro** [mostaphaRoudsari/Honeybee#355]
Hi,
This time i'm trying to evaluate a comfort analysis using the PMVComfRecipe (so far i used the AdaptiveComfort).
The PMV recipe is fine. The HB_IndoorComfAnalysis works fine when the analysis period is one month or two (maybe more). But when i try the whole year this component crashes (turn red) and the readMe output plots the following:
1 JAN 1:00  -  31 DEC 24:00
Runtime error (AggregateException): One or more errors occurred.
Traceback:
  line 1060, in mainPMV, "<string>"
  line 1601, in script
Any ideas? The year analysis took about 40 minutes, it is hard to play with it.
Thanks,
-A.

**Comment by chriswmackey** [comment]
Abraham,
This is my bad.  I changed the PMV function in Ladybug but forgot to commit the new honeybee components.  The version that I just put on the github has worked on my end and, if it does for you too, feel free to close out the issue.  Note that the new commit that I just did involves some name changes for some of the components so you may have to update a few of the comfort map components manually. (https://github.com/mostaphaRoudsari/Honeybee/commit/14b4922b3c4d41169699084d269f5a46723af4d8)

-Chris

**Comment by ayezioro** [comment]
Thanks Chris,
Unfortunately is still not working for the whole year. Same comments on the readMe output of the HB_IndoorComfAnalysis.
Just to be sure the updated components are:
PMVComfRecipe, IndoorViewFactor, readMicroclimateMtx and VisualizeMicroclimate, which are dated Jun_30. The others are dated Jun_21.
-A.

**Comment by ayezioro** [comment]
Forgot to mention that it fails after 22 minutes. Just want to say that it doesn't immediately after starting the run.
-A.

**Comment by chriswmackey** [comment]
@ayezioro ,
The components are working without any issue on my system as you can see by this annual PMV analysis here:
![annualthermalcomfortpercent](https://cloud.githubusercontent.com/assets/5567574/8444623/e0875342-1f60-11e5-96f1-e3974bfa4513.PNG)

and in this gh file:
https://www.dropbox.com/s/a1i47t0b07sqpov/ComfortAutonomyComponent4.gh?dl=0

So the problem seems to be specific to your system.  Also, I just noticed that you did not include the newly-named "Microclimate Map Analysis" in your list so I imagine that this is likely the problem.  I re-named it from its original name "Indoor Comfort Analysis" because the workflow now includes UTCI and an ability to map thermal conditions for the outdoors.

If that component is not the issue, please upload your GH file and I will take a look.  The vague error message that you get is because you are running the component in parallel, which stops me from seeing what the specific problem is in the error message.

-Chris

**Comment by ayezioro** [comment]
Yes!! That was it. Somehow i updated the components before the Map was in the github.
It took about 2 hours (!!). Next time i'll take your suggestion just sending the critical weeks. Otherwise it will be impossible to use.
I'm glad you implemented the possibility of outdoors. Do you have an example for that? Seeing the component, there is a need of building simulation. The question is if this is a must condition or some assumptions can be made for that. I can't imagine you have an urban square and you need to simulate all the buildings around :-)

Thanks,
-A.

**Comment by ayezioro** [comment]
Sorry that i have to re-open the issue. I think i've found a small bug.
When i do a whole year analysis and i connect to the VisualizeMicroclimate a stepOfSimulation input (the HOY) it works fine. But when i do a simulation for one of the period from the LB__statFile the Visualization component fails. It is something like it expects the simulation starts at 1st of January (or something like this), setting the HOY starting from that shows results.
Can be?
-A.

**Comment by chriswmackey** [comment]
@ayezioro ,
To give you a sense of what is taking so much time in your run, almost all of it is going to the calculation of PMV and not the zone air and radiant temperature.  So your options for a faster simulation include:

1) USE ADAPTIVE COMFORT - In my experience, running an adaptive recipe usually takes about 1/8th of the time than the same run with a PMV recipe.  The calculation of PMV is incredibly complicated and involves a whole calculation of heat balance over an hour or until equilibrium is reached (this is where your calculation time is going).  PMV is way more computationally intensive than it needs to be and much new research is showing that it is not even as good of an indicator of comfort as the much simpler adaptive model.  Personally, I have been looking to take PMV completely out of my workflow for a while and the next stable LB release will allow users to make use of new research that has applied the adaptive methodology of surveying people in real buildings to conditioned cases (not just naturally ventilated ones).  With this, I am going to recommend that people use the PMV model at the end of a design process to show code compliance and that they are better off using the adaptive model with more regular feedback during a design process.

2) INCREASE YOUR GRID SIZE - I have found in my research so far that, if you are mostly concerned about it totaling the comfortable hours, you don't need a grid size much finer than the space that an individual person would occupy.  So a 1 meter grid size has given me comparable results to a finer grid size when you look at averages over an entire room.  Of course, if you are looking to make cool visualizations and animations, the finer grid size is nice but, for these, you often do not need as high of a temporal resolution. And this brings us to the last option:

3) RUN THE MODEL FOR A SHORT ANALYSIS PERIOD - As you already stated, if you are only concerned about what happens in extreme or typical cases, then you only need to run the model for an extreme or typical week or an average week of July, etc.


To the outdoor question: the buildings around an urban space (especially a small urban space) are very important to understanding its micro-climate.  Particularly, the question of whether these buildings have massive or insulated envelopes and whether they are conditioned or unconditioned has a huge impact on the temperature felt within the outdoor space around them.

For example, the cooling effect of very narrow street canyons seen in many traditional middle-eastern cities is almost entirely due to the effect of shaded thermal mass of the surrounding buildings.  This cooling effect of narrow streets goes against a lot of traditional urban heat island research, which states that these narrow streets should increase the heat island because they stop daytime heat from re-radiating back to the sky at night.  While this re-radiating principle works for streets with height-to-with ratios around 1, I know that it falls apart at much narrower levels because many of these traditional UHI models do not adequately account for the effect of the building's shaded thermal mass (which can diminish UHI if daytime heat is night flushed) or the fact that neighboring buildings can be conditioned (which can diminish the street-level UHI by removing heat through rooftop mechanical systems).  These traditional studies also often obsess way too much about the air temperature of UHI instead of looking at the radiant temperature of these surrounding (potentially shaded and  thermally massive) surfaces.  So part of the reason for me putting in this outdoor option is a reaction that I have had against this traditional line of research, which I would like to critique and inform.  Here is an example file where you can play around with the narrowness of a street canyon and see the effect on micro-climate:
https://app.box.com/s/zds0g23834360i3ha0dbnrrj3ofzr4t2

I am hoping that I can turn this into a paper at some point soon.
-Chris

**Comment by chriswmackey** [comment]
@ayezioro ,
In response to you re-opening the issue, I am not experiencing the same situation on my machine (the stepOfSimulation is always starting from 0).  Perhaps it is a specific case that I did not catch.  Can you post a link to your GH file?
-Chris

**Comment by ayezioro** [comment]
Great explanation. You should write the paper!!
Attached the file i'm working with. Before you open it, set the tolerance to 0.001 (!!)
See red marked zone at the right end of the file.
Say the extreme week happens between 3.1 through 3.7. In the DOY/HOY i set the day/month accordingly and then i get the error (i hook the HOY output). If i set the month to 1 and the day to 1 i get the result.
This is the link to the file: https://www.dropbox.com/s/donad8o8wh0ttx4/AngelHouse.gh?dl=0
Thanks,
-A.

**Comment by ayezioro** [comment]
I know this is not the proper issue to ask questions about the OutdoorComfRecipe. How you take into consideration the outdoor wind speed and relative humidity (as you do, for instance, in the LB_OutdoorComfortCalculator)? I see that you take those from the building simulation, which are indoors.
Another question:
I can understand that you take the surfaces outdoor temperatures, but why you need the indoor temp, _zoneAirFlowVol and _zoneAirHeatGain?
I should probably open a new issue or discussion ...
Thanks,
-A.

**Comment by chriswmackey** [comment]
@ayezioro ,
Thanks for the encouragement.

I should have realized this sooner by re-reading your explanation but the issue that you bring up is not a bug but just a disagreement about how the "stepOfSimulation" should work.  On both the comfort maps and the Color Zones/ Color Surfaces components, the stepOfSimulation refers to the step of whatever analysis was run and not the HOY.  For the Color Zones/ Color Surfaces, this was necessary because the simulation data is sometimes monthly, daily, or at the simulation timestep (not hourly).  I can change the input on the comfort map components to actually be an HOY since the analysis currently only works for hourly timesteps.  However, I was hoping at some point down the line to put in functionality to allow users to make thermal maps at the E+ timestep if people really need a high temporal resolution for animations.  As a result, I would rather keep the stepOfSimulation the way it is unless you know if this has caused confusion for other users.

I only imagined people using the stepOfSimulation input to make animations and thought that, if people were really interested in a specific hour of the year, they would use the analysisPeriod input.  Also, I really recommend using the analysis period to take out all of the hours of a specific time of day for a simulation that you have run like so:
![analysisperiodhourslice](https://cloud.githubusercontent.com/assets/5567574/8470607/fe866b84-205b-11e5-9465-aedb01a86b22.PNG)


This averaging of the conditions across a time period helps weed out weird anomalies of specific hours to see the broader trends.  This is also how I created these animations:
https://www.youtube.com/watch?v=WJz1Eojph8E&index=3&list=PLruLh1AdY-Sj3ehUTSfKa1IHPSiuJU52A
https://www.youtube.com/watch?v=CATJN-rOkNM&index=4&list=PLruLh1AdY-Sj3ehUTSfKa1IHPSiuJU52A

If you are satisfied with the explanation, feel free to close out the issue.  Otherwise, we can discuss.
-Chris

**Comment by chriswmackey** [comment]
To answer the questions about the outdoors, I am going to put out full written and video documentation for how the maps work once I put in this last bit of functionality to calculate thermal autonomy.

I should stress that the outdoor maps are not a complete representation of all outdoor conditions but they represent the assumptions about the outdoor environment made by E+ and they are the best that we can do at this point in time.  The outdoor air temperatures, humidities, and wind speeds are all taken from the epwFile input to the UTCI analysis recipe.  All outdoor air is assumed to be well mixed in terms of temperature and, In the case of wind, I use the power law that E+ uses.  So there are no horizontal changes in wind speed (only vertical).  Of course, in reality, the air temperature + wind speed in the outdoors is much more complex to the point that even modern-day CFD studies raise doubts and this is why there are many people making PhDs and careers out of this outdoor air challenge.  For the time being, we can trust these outdoor maps to create a decent representation of the radiant environment (which seems to be a larger player) and accept a margin of error in the air environment (particularly with the wind speed).

Feel free to open another discussion if you still have questions,
-Chris

**Comment by ayezioro** [comment]
Closing it. Thank you.
As for the outdoors, i was sure you are taking the epw in some way. Just wanted to confirm :-)
Just one more little one: the _zoneAirFlowVol and _zoneAirHeatGain are not needed for this recipe, right? If so, maybe it will be better to take them out, or at least not making them compulsory?
-A.

**Comment by chriswmackey** [comment]
@ayezioro ,

All three of the comfort map recipes currently support both indoor and outdoor thermal maps regardless of whether the comfort model is developed for use in indoor or outdoor conditions.  I do this because I often felt compelled to compare the indoor and outdoor conditions to one another (particularly with semi-outdoor spaces like balconies and courtyards) and because I trust that users who are advanced enough to build these maps will know the conditions that each of the maps were designed for and not abuse this.

You can create comfort maps for both indoor and outdoor conditions simultaneously just by putting a surface that spans both interior and exterior of the zones into the distFromFlrOrSrf_ of the view factor component and setting includeOutdoor to 'true' (the default of includeOutdoor is set to False and this will automatically exclude any part of an input surface that is outdoors). In this way, all of the comfort maps follow the philosophy that the maps are really made to describe indoor conditions and the ability to show the outdoors is just an extra bonus that should be used with caution (for previously-stated reasons in this issue).

I followed this philosophy even with the UTCI model because EnergyPlus is primarily made to describe indoor conditions but I can understand the desire to make the UTCI model with optional inputs for indoor conditions.  It will involve some re-structuring of the code that will take some time on my end but I see how this will save time in the long run by having E+ not write out certain outputs.  I have created an issue to take care of this in the future:

https://github.com/mostaphaRoudsari/Honeybee/issues/357

**Comment by ayezioro** [comment]
This is awesome!!
I have no doubt it will be abused :-) but this opens opportunities which is great.
Your upcoming videos will be an invaluable source ;-)
-A.

**Comment by chriswmackey** [comment]
@ayezioro ,
The indoor variables have been made optional for the UTCI map with this:
https://github.com/mostaphaRoudsari/Honeybee/commit/768afbe32393916bd2364482d32b1fe481d6b238
https://github.com/mostaphaRoudsari/Honeybee/issues/357
-Chris

**Comment by ayezioro** [comment]
Great!!
Thanks.
-A.
